### PR TITLE
Implement proper error handling for Rx types.

### PR DESCRIPTION
**Pull request by JakeWharton**
Since we've implemented our own observables instead of using the built-in factories we have to do the heavy lifting of these edge cases ourselves.

### CODE REVIEW

File: retrofit-adapters/rxjava/src/test/java/retrofit2/adapter/rxjava/CompletableTest.java [hunk 1]
```diff
@@ -69,4 +76,56 @@
...
   }
+
+  @Test public void bodyThrowingInOnCompletedDeliveredToPlugin() {
+    server.enqueue(new MockResponse().setBody("Hi"));
+
+    final AtomicReference<Throwable> throwableRef = new AtomicReference<>();
+    RxJavaPlugins.getInstance().registerErrorHandler(new RxJavaErrorHandler() {
+      @Override public void handleError(Throwable throwable) {
+        if (!throwableRef.compareAndSet(null, throwable)) {
+          throw Exceptions.propagate(throwable);
```
Comment (swankjesse) [inline retrofit-adapters/rxjava/src/test/java/retrofit2/adapter/rxjava/CompletableTest.java]: nice

File: retrofit-adapters/rxjava/src/test/java/retrofit2/adapter/rxjava/CompletableTest.java [hunk 2]
```diff
@@ -16,20 +16,27 @@
...
+import rx.exceptions.Exceptions;
+import rx.plugins.RxJavaErrorHandler;
+import rx.plugins.RxJavaPlugins;
 
 import static okhttp3.mockwebserver.SocketPolicy.DISCONNECT_AFTER_REQUEST;
 import static org.assertj.core.api.Assertions.assertThat;
 
 public final class CompletableTest {
   @Rule public final MockWebServer server = new MockWebServer();
+  @Rule public final TestRule pluginsReset = new RxJavaPluginsResetRule();
```
Comment (swankjesse) [inline retrofit-adapters/rxjava/src/test/java/retrofit2/adapter/rxjava/CompletableTest.java]: these are so much better than setUp/tearDown

File: retrofit-adapters/rxjava/src/test/java/retrofit2/adapter/rxjava/ObservableTest.java [hunk 3]
```diff
@@ -171,4 +312,73 @@
...
+        }
+      }
+    });
+
+    RecordingSubscriber<Result<String>> subscriber = subscriberRule.create();
+    final RuntimeException first = new RuntimeException();
+    final RuntimeException second = new RuntimeException();
+    service.result().unsafeSubscribe(new ForwardingSubscriber<Result<String>>(subscriber) {
+      @Override public void onNext(Result<String> value) {
+        // The only way to trigger onError for a result is if onNext throws.
```
Comment (swankjesse) [inline retrofit-adapters/rxjava/src/test/java/retrofit2/adapter/rxjava/ObservableTest.java]: sneaky

File: retrofit-adapters/rxjava/src/test/java/retrofit2/adapter/rxjava/SingleTest.java [hunk 4]
```diff
@@ -126,4 +268,74 @@
...
+      @Override public void onCompleted() {
+        throw e;
+      }
+    });
+
+    assertThat(subscriber.takeValue().response().body()).isEqualTo("Hi");
+    assertThat(throwableRef.get()).isSameAs(e);
+  }
+
+  @Ignore("Single's contract is onNext|onError so we have no way of triggering this case")
```
Comment (swankjesse) [inline retrofit-adapters/rxjava/src/test/java/retrofit2/adapter/rxjava/SingleTest.java]: interesting

**Comment by swankjesse** [comment]
Tests are repetitive, but everything makes sense.
